### PR TITLE
[feat] Add an SSH-based scheduler backend

### PR DESCRIPTION
**Pull request by vkarak**
The new backend is called `ssh` and will spawn a test job on a remote host. More specifically, it will copy the test's stage directory to the remote host, execute the job and copy back the results.

Multiple hosts can be assigned to a partition with an `ssh` scheduler, in which case the scheduler will pick the next free host to spawn a test on.

This PR also introduced a new `run_command_async2` utility for executing asynchronous commands. Conversely to the `run_command_async`, this returns a "future" encapsulating the spawned process. Process futures can be chained with the `then()` method and the chained futures will be called upon termination of the predecessor in the chain.

The SSH scheduler backend uses the process futures and spawns a chain of push artefacts-execute job-pull artefacts commands in its `submit()` method, so that it does not block. The progress is ensured by the scheduler's `poll()` method which polls the spawned futures for completion.

This PR essentially solves the feature request in #64.

More details about the new scheduler can be found in its docs.

### CODE REVIEW

File: reframe/core/schedulers/ssh.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        job.steps['exec'] = osext.run_command_async2(
+            f'ssh -o BatchMode=yes {options} {job.host} '
+            f'"cd {job.remotedir} && bash -l {job.script_filename}"'
+        )
+
+    def submit(self, job):
+        assert isinstance(job, _SSHJob)
+
+        # Check if `#host` pseudo-option is specified and use this as a host,
+        # stripping it off the rest of the options
```
Comment (vkarak) [inline reframe/core/schedulers/ssh.py:131]: Forgot to document this.

File: reframe/core/schedulers/ssh.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        host = None
+        stripped_opts = []
+        options = job.sched_access + job.options + job.cli_options
+        for opt in options:
+            if opt.startswith('#host='):
+                _, host = opt.split('=', maxsplit=1)
+            else:
+                stripped_opts.append(opt)
+
+        # Host is pinned externally (`--distribute` option)
```
Comment (vkarak) [inline reframe/core/schedulers/ssh.py:141]: A note about using `--distribute` with this backend is worth in the docs, too.

File: reframe/schemas/config.json [hunk 3]
```diff
@@ -109,6 +109,10 @@
         "sched_options": {
             "type": "object",
             "properties": {
+                "hosts": {
```
Comment (vkarak) [inline reframe/schemas/config.json:112]: Should I change this to `ssh_hosts`, wdyt?

File: reframe/core/schedulers/ssh.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    def host(self):
+        return self._host
+
+    @property
+    def ssh_options(self):
+        return self._ssh_options
+
+
+@register_scheduler('ssh')
+class SSHJobScheduler(JobScheduler):
```
Comment (vkarak) [inline reframe/core/schedulers/ssh.py:45]: I would also need to add some basic unit tests for this scheduler, too.

**Review by victorusu — APPROVED** [review]
`lgtm`. I wonder how we will do the unit tests for this scheduler. Do we have any plans?

**Comment by vkarak** [comment]
> lgtm. I wonder how we will do the unit tests for this scheduler. Do we have any plans?

I just added some basic unit tests.

**Pull request merged**